### PR TITLE
overlord/ifacestate: fix migration of connections on upgrade from ubuntu-core

### PR DESCRIPTION
**Pull request by stolowski**
The connections migration code for ubuntu-core -> core transition migrates connections stored in "conns" from ubuntu-core to core and reloads connections in the repo (via reloadConnections helper). It hower incorrectly calls reloadConnections for "ubuntu-core" snap under the assumption that reloadConnections takes care of full synchronization of conns -> repo and removes non-existing connections, which is not the case. So as a result we end up with correct "conns" but incorrect repo with for both "core" and "ubuntu-core" connection. This causes an error when we run "auto-disconnect" task on "ubuntu-core" removal - see https://paste.ubuntu.com/p/FcK6mx2SwZ/
The spread test updated with this PR uncovers the problem and reports the failure of migration without the fix (thanks Michael).

The fix explicitely removed connections of "ubuntu-core" from the repo as part of migration task, so that "conns" and "repo" are in sync.

### CODE REVIEW

File: overlord/ifacestate/handlers.go [hunk 1]
```diff
@@ -1135,14 +1135,17 @@ func (m *InterfaceManager) transitionConnectionsCoreMigration(st *state.State, o
 	}
 	setConns(st, conns)
 
+	// After migrating connections in state, remove them from repo so they stay in sync and we don't
```
Comment (zyga) [inline overlord/ifacestate/handlers.go]: The `ubuntu-core` snap used to have a configuration hook and said hook had a plug for `core-support`. I think we must remove that plug connection as well.

File: overlord/ifacestate/handlers.go [hunk 2]
```diff
@@ -1135,14 +1135,17 @@ func (m *InterfaceManager) transitionConnectionsCoreMigration(st *state.State, o
...
+	// attempt to run disconnects on when the old core gets removed as part of the transition.
+	if err := m.removeConnections(oldName); err != nil {
+		return err
+	}
+
 	// The reloadConnections() just modifies the repository object, it
 	// has no effect on the running system, i.e. no security profiles
 	// on disk are rewriten. This is ok because core/ubuntu-core have
 	// exactly the same profiles and nothing in the generated policies
 	// has the slot-name encoded.
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: this preexisting comment is a bit confusing. What about snap-confine profiles? also I think s/slot-name/core snap-name/ ?

File: tests/main/upgrade-from-2.15/task.yaml [hunk 3]
```diff
@@ -38,3 +38,13 @@ execute: |
...
+
+    echo "wait for ubuntu-core->core transition"
+    snap debug ensure-state-soon
+    for i in $(seq 240); do
+        if snap changes|grep -q ".*Done.*Transition ubuntu-core to core" ; then
+            break
+        fi
+        sleep 1
+    done
+    snap changes|MATCH ".*Done.*Transition ubuntu-core to core"
```
Comment (pedronis) [inline tests/main/upgrade-from-2.15/task.yaml]: we should probably check the status of the snap-confine profiles here

**Comment by zyga** [comment]
Approved because ubuntu-core never had the `core-support` plug I was thinking about.

File: tests/main/upgrade-from-2.15/task.yaml [hunk 4]
```diff
@@ -38,3 +38,19 @@ execute: |
...
+  
+    echo "check that snap confine profiles are fine after upgrade"
```
Comment (zyga) [inline tests/main/upgrade-from-2.15/task.yaml]: This looks good! Thank you for adding that.

File: tests/main/upgrade-from-2.15/task.yaml [hunk 5]
```diff
@@ -33,8 +33,30 @@ execute: |
     fi
     wait_for_service snap.go-example-webserver.webserver.service
 
+    echo "Install a test snap"
+    snap install hello-world
```
Comment (mvo5) [inline tests/main/upgrade-from-2.15/task.yaml]: (nitpick) Hey, thanks for adding this smoke test! Is there a downside in using `test-snapd-tools` here? It's slightly more common for this kind of test. OTOH now that the store does not count installs from integration tests for the metrics it probably does not matter I much guess(?).

**Comment by pedronis** [comment]
@stolowski @mvo5 there is a shellcheck problem with the new test, also we get archive Hash Sum issues on xenial :/

**Comment by mvo5** [comment]
Thanks @pedronis - I can push a trivial fix in 5min (just need to run spread once to double check its good):
```
index cd444c568..98e7fe369 100644
--- a/tests/main/upgrade-from-2.15/task.yaml
+++ b/tests/main/upgrade-from-2.15/task.yaml
@@ -34,7 +34,7 @@ execute: |
     wait_for_service snap.go-example-webserver.webserver.service
 
     echo "Install a test snap"
-    snap install hello-world
+    snap install test-snapd-tools.echo
     
     echo "upgrade to current snapd"
     apt install -y "$GOHOME"/snapd*.deb
@@ -54,9 +54,9 @@ execute: |
   
     echo "check that snap confine profiles are fine after upgrade"
     test -f /etc/apparmor.d/usr.lib.snapd.snap-confine.real
-    test $(ls /var/lib/snapd/apparmor/profiles/snap-confine.core.* 2>/dev/null | wc -l) -eq 1
+    test "$(find /var/lib/snapd/apparmor/profiles/ -name "snap-confine.core.*" | wc -l)" -eq 1
     MATCH "^/usr/lib/snapd/snap-confine \\(enforce\\)" < /sys/kernel/security/apparmor/profiles
     MATCH "^/snap/core/.*/usr/lib/snapd/snap-confine \\(enforce\\)" < /sys/kernel/security/apparmor/profiles
 
     echo "Smoke test, this should fail if profiles were wrong"
-    hello-world
+    test-snapd-tools.echo hello | MATCH hello
```

**Comment by pedronis** [comment]
@mvo5 thanks,    

     +    snap install test-snapd-tools.echo     

no ".echo"  needed there :)

File: tests/main/upgrade-from-2.15/task.yaml (hunk 5 above)
Comment (stolowski) [reply tests/main/upgrade-from-2.15/task.yaml]: Yes, makes sense, thanks for pushing a change.

**Comment by stolowski** [comment]
> +    test "$(find /var/lib/snapd/apparmor/profiles/ -name "snap-confine.core.*" | wc -l)" -eq 1

@mvo5 Ah, right, thanks for the fix!

**Comment by mvo5** [comment]
@stolowski My pleasure :) Lets hope it passes and then we can squash and cherry-pick and do the 2.37.4 release.

**Pull request merged**